File: daily/2026/1594_Maximum_Non_Negative_Product_in_a_Matrix.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProductPath(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<vector<long long>>> dp(m, vector<vector<long long>>(n, vector<long long>(2, -1)));
        if (grid[0][0] > 0) dp[0][0][0] = grid[0][0];
        else if (grid[0][0] < 0) dp[0][0][1] = -grid[0][0];
        else dp[0][0][0] = dp[0][0][1] = 0;
        for (int j = 1; j < n; j++){
            if (grid[0][j] > 0){
                if (dp[0][j-1][0] != -1) dp[0][j][0] = dp[0][j-1][0] * grid[0][j];
                if (dp[0][j-1][1] != -1) dp[0][j][1] = dp[0][j-1][1] * grid[0][j];
            }
            else if (grid[0][j] < 0){
                if (dp[0][j-1][0] != -1) dp[0][j][1] = dp[0][j-1][0] * -grid[0][j];
                if (dp[0][j-1][1] != -1) dp[0][j][0] = dp[0][j-1][1] * -grid[0][j];
            }
            else{
                dp[0][j][0] = 0;
                dp[0][j][1] = 0;
            }
        }
        for (int i = 1; i < m; i++){
            if (grid[i][0] > 0){
                if (dp[i-1][0][0] != -1) dp[i][0][0] = dp[i-1][0][0] * grid[i][0];
                if (dp[i-1][0][1] != -1) dp[i][0][1] = dp[i-1][0][1] * grid[i][0];
            }
            else if (grid[i][0] < 0){
                if (dp[i-1][0][0] != -1) dp[i][0][1] = dp[i-1][0][0] * -grid[i][0];
                if (dp[i-1][0][1] != -1) dp[i][0][0] = dp[i-1][0][1] * -grid[i][0];
            }
            else{
                dp[i][0][0] = 0;
                dp[i][0][1] = 0;
            }
        }
        for (int i = 1; i < m; i++){
            for (int j = 1; j < n; j++){
                if (grid[i][j] > 0){
                    if (max(dp[i-1][j][0], dp[i][j-1][0]) != -1) dp[i][j][0] = max(dp[i-1][j][0], dp[i][j-1][0]) * grid[i][j];
                    if (max(dp[i-1][j][1], dp[i][j-1][1]) != -1) dp[i][j][1] = max(dp[i-1][j][1], dp[i][j-1][1]) * grid[i][j];
                }
                else if (grid[i][j] < 0){
                    if (max(dp[i-1][j][1], dp[i][j-1][1]) != -1) dp[i][j][0] = max(dp[i-1][j][1], dp[i][j-1][1]) * -grid[i][j];
                    if (max(dp[i-1][j][0], dp[i][j-1][0]) != -1) dp[i][j][1] = max(dp[i-1][j][0], dp[i][j-1][0]) * -grid[i][j];
                }
                else{
                    dp[i][j][0] = 0;
                    dp[i][j][1] = 0;
                }
            }
        }
        if (dp[m-1][n-1][0] == -1) return -1;
        return dp[m-1][n-1][0] % (long long)(1e9 + 7);
    }
};

int main(){
    vector<vector<int>> grid = {{-1,-2,-3},{-2,-3,-3},{-3,-3,-2}};
    // vector<vector<int>> grid = {{1,-2,1},{1,-2,1},{3,-4,1}};
    // vector<vector<int>> grid = {{1,3},{0,-4}};
    Solution S;
    cout << S.maxProductPath(grid) << endl;
    return 0;
}

File: daily_python/1594_Maximum_Non_Negative_Product_in_a_Matrix.py
import os
import sys
from typing import List

class Solution:
    def maxProductPath(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        Maxp = [[0] * n for _ in range(m)]
        minp = [[0] * n for _ in range(m)]
        Maxp[0][0] = minp[0][0] = grid[0][0]
        for i in range(1, m):
            Maxp[i][0] = minp[i][0] = Maxp[i-1][0] * grid[i][0]
        for j in range(1, n):
            Maxp[0][j] = minp[0][j] = Maxp[0][j-1] * grid[0][j]
        for i in range(1, m):
            for j in range(1, n):
                if grid[i][j] >= 0:
                    Maxp[i][j] = max(Maxp[i-1][j], Maxp[i][j-1]) * grid[i][j]
                    minp[i][j] = min(minp[i-1][j], minp[i][j-1]) * grid[i][j]
                else:
                    Maxp[i][j] = min(minp[i-1][j], minp[i][j-1]) * grid[i][j]
                    minp[i][j] = max(Maxp[i-1][j], Maxp[i][j-1]) * grid[i][j]
        if Maxp[m-1][n-1] < 0:
            return -1
        return Maxp[m-1][n-1] % (10 ** 9 + 7)
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            grid = [list(map(int, row.split(','))) for row in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append(grid)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for grid in dataset:
        results.append(solution.maxProductPath(grid))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')